packages/omezarr/src/zarr/loading.ts

```diff
@@ -0,0 +1,326 @@
+}
+
+// feel free to freak out if the request is over or under determined or whatever
+function buildQuery(r: Readonly<ZarrRequest>, axes: readonly OmeZarrAxis[], shape: readonly number[]) {
+    const ordered = axes.map((a) => r[a.name as ZarrDimension]);
+    // if any are undefined, throw up
+    if (ordered.some((a) => a === undefined)) {
+        throw new Error('request does not match expected dimensions of ome-zarr dataset!');
+    }
+
+    return ordered.map((d, i) => {
+        const bounds = { min: 0, max: shape[i] };
+        if (d === null) {
+            return d;
+        }
+        if (typeof d === 'number') {
+            return limit(bounds, d);
+        }
+        return zarr.slice(limit(bounds, d.min), limit(bounds, d.max));
+    });
+}
+
+export async function explain(z: OmeZarrMetadata) {
+    logger.dir(z);
+}
+
+export function indexOfDimension(axes: readonly OmeZarrAxis[], dim: ZarrDimension) {
+    return axes.findIndex((ax) => ax.name === dim);
+}
+/**
+ * get voxels / pixels from a region of a layer of an omezarr dataset
+ * @param metadata a ZarrMetadata from which to request a slice of voxels
+ * @param r a slice object, describing the requested region of data - note that it is quite possible to request
+ * data that is not "just" a slice. The semantics of this slice object should match up with conventions in numpy or other multidimensional array tools:
+ * @see https://zarrita.dev/slicing.html
+ * @param level the layer within the LOD pyramid of the OME-Zarr dataset.
+ * @returns the requested chunk of image data from the given layer of the omezarr LOD pyramid. Note that if the given layerIndex is invalid, it will be treated as though it is the highest index possible.
+ * @throws an error if the request results in anything of lower-or-equal dimensionality than a single value
+ */
+export async function loadSlice(metadata: OmeZarrMetadata, r: ZarrRequest, level: OmeZarrShapedDataset) {
+    // put the request in native order
+    const store = new zarr.FetchStore(metadata.url);
+    const scene = metadata.attrs.multiscales[0];
+    const { axes } = scene;
+    if (!level) {
+        const message = 'invalid Zarr data: no datasets found';
+        logger.error(message);
+        throw new VisZarrDataError(message);
+    }
+    const arr = metadata.arrays.find((a) => a.path === level.path);
+    if (!arr) {
+        const message = `cannot load slice: no array found for path [${level.path}]`;
+        logger.error(message);
+        throw new VisZarrDataError(message);
+    }
+    const { raw } = await loadZarrArrayFileFromStore(store, arr.path, 2, false);
+    const result = await zarr.get(raw, buildQuery(r, axes, level.shape));
+    if (typeof result === 'number') {
+        throw new Error('oh noes, slice came back all weird');
+    }
+    return {
+        shape: result.shape,
+        buffer: result,
+    };
+}
+
+export async function load(url: string) {
```

Comment:
Here's the essentially identical export that I mentioned in another comment to consolidate.